### DISCUSSION TITLE
Attempting to edit the asreview variables after they have been created

### DISCUSSION THREAD

**Discussion by michagast** [asreview/asreview#1228] [Troubleshooting (report errors in the "Issues" tracker)]
Hi everyone,

For my masters thesis I am using active learning to perform text classification. I'm using ASReview for the active learning part of this by extracting the models it creates and then applying these on data outside of ASReview to see how well they can predict the labels of new data. So far my results have been promising, as I have been able to show that using active learning improves the learning speed of models by using the simulation mode, especially that of logistic regression. To verify my results that I have gotten so far even more, I want to test the workings of asreview using a holdout set. To do this I have modified the asreview base code slightly. This is because one of the features of my model, namely a bag of words, needs to be taken over the entire dataset, including the holdoutset. To account for this feature, I remove the indices that are in the holdoutset after I have created the feature matrix(which is created using my of feature extraction extenstion [link](https://github.com/michagast/asreview) ).

So far I have been unable to generate logical results from ASReview with this holdout set. I am struggling to see which specific parts of the code of ASReview in need to adjusted in order to make sure it properly works with this holdout set. The adjustments i made to review/base.py can be seen below:

```
  def __init__(
        self,
        as_data,
        project,
        model=NaiveBayesClassifier(),
        query_model=MaxQuery(),
        balance_model=SimpleBalance(),
        feature_model=Tfidf(),
        n_papers=None,
        n_instances=DEFAULT_N_INSTANCES,
        stop_if=None,
        start_idx=[],
    ):
        """Initialize the reviewer base class, so that everything is ready to
        train a new model."""
        super(BaseReview, self).__init__()

        # Set the model.
        self.classifier = model
        self.balance_model = balance_model
        self.query_strategy = query_model
        self.feature_extraction = feature_model

        # Set the settings.
        self.as_data = as_data
        self.project = project
        self.n_instances = n_instances
        self.stop_if = stop_if
        self.prior_indices = start_idx
        # Import the holdout indices and take the the first column as index since this is where the proper values are
        holdout_data = pd.read_excel(r'enron_holdoutdata.xlsx', index_col=[0])
        if n_papers is not None:
            logging.warning("Argument n_papers is deprecated, ignoring n_papers.")

        # Get the known labels.
        self.data_labels = as_data.labels
        if self.data_labels is None:
            self.data_labels = np.full(len(as_data), LABEL_NA)

        with open_state(self.project, read_only=False) as state:
            # If the state is empty, add the settings.
            if state.is_empty():
                state.settings = self.settings
            self.record_table = state.get_record_table()

            # Add the record table to the state if it is not already there.
            #self.record_table = state.get_record_table()
            #if self.record_table.empty:
                #state.add_record_table(as_data.record_ids)
                #self.record_table = state.get_record_table()

            # Retrieve feature matrix from the project file or create
            # one from scratch.
            try:
                self.X = self.project.get_feature_matrix(
                    self.feature_extraction.name)
            except FileNotFoundError:
                self.X = self.feature_extraction.fit_transform(
                    as_data.texts,
                    as_data.headings,
                    as_data.bodies,
                    as_data.keywords
                )

                # remove holdout indexes from the as_data dataframe
                as_data.df.drop(holdout_data.index.values, axis=0, inplace=True)
                
                # add record ids now since now the holdout indices have been removed
                state.add_record_table(as_data.record_ids) 
                self.record_table = state.get_record_table()
                if self.X.shape[0] != len(as_data.df):
                    raise ValueError(
                        "Dataset has {} records while feature "
                        "extractor returns {} records"
                        .format(len(as_data), self.X.shape[0]))

                self.project.add_feature_matrix(self.X,
                                                self.feature_extraction.name)

            # remove the label indices that are part of the holdout set
            self.data_labels = np.delete(self.data_labels, holdout_data.index.values, axis=0)
            # Check if the number of records in the feature matrix matches the
            # length of the dataset.
            if self.X.shape[0] != len(self.data_labels):
                raise ValueError("The state file does not correspond to the "
                                 "given data file, please use another state "
                                 "file or dataset.")
            # update asreview settings
            state.settings = self.settings

            # Make sure the priors are labeled.
            self._label_priors()
            state.settings = self.settings
```

Furthermore I made adjustments to data/base.py since I don't reset the index of the dataframe and therefore I want the record function to use loc instead of iloc:
```

    def record(self, i, by_index=False):
        """Create a record from an index.

        Arguments
        ---------
        i: int, iterable
            Index of the record, or list of indices.
        by_index: bool
            If True, take the i-th value as used internally by the review.
            If False, take the record with record_id==i.

        Returns
        -------
        PaperRecord
            The corresponding record if i was an integer, or a list of records
            if i was an iterable.
        """
        if not is_iterable(i):
            index_list = [i]
        else:
            index_list = i

        if by_index:
            records = [
                PaperRecord(**self.df.iloc[j],
                            column_spec=self.column_spec,
                            record_id=self.df.index.values[j])
                for j in index_list
            ]
        else:
            records = [
                PaperRecord(**self.df.loc[j, :],
                            record_id=j,
                            column_spec=self.column_spec) for j in index_list
            ]

        if is_iterable(i):
            return records
        return records[0]
```

I really hope that there is someone who can help, as the results that i have gotten so far seem very promising and by using this holdout set the results could be validated properly.

One weird type of behaviour that is introduced by these changes is that at one point the labeled set never changes anymore. I have been trying to figure out what causes this but I have not been able to find it.

Thanks in advance for any help or comments, they are both greatly appreciated.

Micha

**Comment by PeterLombaers** [comment]
Hi Micha,

Deleting rows in the feature matrix might result in a mismatch between the records in your dataset and the corresponding vectors. Trying to change how this works in review/base.py and other places (you would probably need to be in review/simulate.py as well!) might be possible, but seems very difficult and prone to errors. 

Is there no way to take out your holdout set directly in the code that generates the feature matrix? Maybe you can pre-calculate the feature matrices using a separate script, and then let ASReview use this precalculated feature matrix? Or you pre-calculate the BOW and use a standard ASReview feature extracter, but set the BOW to the pre-calculated value? That would seem a lot less likely to get errors because you are not changing the workings of the whole program. It would also be a lot easier for others to reproduce your work: they would just need the generated files and the feature_extractor extension, instead of a whole different version of ASReview.

Peter
